### PR TITLE
All-around improvements to toolbox loading/reloading

### PR DESCRIPTION
**Pull request by mvdbeek**
- Fix hang when installing data managers
- Finish reloading the new toolbox before replacing the old one
- Retry parsing of tool conf file when a xml parsing errors occurs
- Log the time it takes to reload the toolbox
- Speed up toolbox and panel loading by caching database requests as much as possible. Speedup is 20-40x (reloading the toolbox went from 30 seconds to 1)

### CODE REVIEW

**Comment by blankenberg** [comment]
@galaxybot test this

**Comment by martenson** [comment]
@galaxybot test this

**Comment by mvdbeek** [comment]
Hmm, both failed tests pass locally. I'll try again.

@galaxybot test this

**Comment by martenson** [comment]
@galaxybot test this

**Comment by martenson** [comment]
@mvdbeek it has to be a solo comment, sorry :/

**Comment by mvdbeek** [comment]
Now it's a different test that fails. Seems something is a bit fragile in 1430 and 1090 :/.

**Comment by mvdbeek** [comment]
@galaxybot test this

File: lib/galaxy/util/profiling.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+http://docs.sqlalchemy.org/en/latest/faq/performance.html#code-profiling
+"""
+import contextlib
+import cProfile
+import pstats
+import StringIO
+
+
+@contextlib.contextmanager
+def profiled():
```
Comment (martenson) [inline lib/galaxy/util/profiling.py]: is this used somewhere or just preparation for future?

File: lib/galaxy/tools/toolbox/lineages/tool_shed.py [hunk 2]
```diff
@@ -7,6 +7,23 @@
     ToolVersion = None
 
 
+class ToolVersionCache(object):
+    def __init__(self, app):
```
Comment (martenson) [inline lib/galaxy/tools/toolbox/lineages/tool_shed.py:18]: Could you please add a bit of a documentation here and below?

**Comment by martenson** [comment]
I like the version cache. 👍 

I was also thinking I would like to have the 'large scale test' you described [here](https://github.com/galaxyproject/galaxy/pull/2840#issuecomment-243820357) as a part of https://github.com/galaxyproject/galaxy/issues/2642 - do you think it possible/useful @mvdbeek @jmchilton ?

**Comment by mvdbeek** [comment]
Sure, I already did it midway through this cycle, was cutting the install time in half

File: lib/galaxy/util/profiling.py (hunk 1 above)
Comment (mvdbeek) [reply lib/galaxy/util/profiling.py]: I did use it to figure out what's slow, I'll remove it (and maybe put that on pypi if I can't find something there, it's quite handy).

Comment (martenson) [reply lib/galaxy/util/profiling.py]: I will definitely use it, thanks for sharing!

**Comment by martenson** [comment]
@mvdbeek I am not sure I understand. I meant if it is feasible to create a def.yml/Dockerfile etc. as in https://github.com/jmchilton/galaxy-community-qa/tree/master/recipes to test installing usegalaxy's toolset in a multi-process environment.

**Comment by jmchilton** [comment]
This is absolutely wonderful, a million thanks @mvdbeek!

Enthusiastic :+1: from me.

**Comment by mvdbeek** [comment]
> @mvdbeek I am not sure I understand. I meant if it is feasible to create a def.yml/Dockerfile etc. as in https://github.com/jmchilton/galaxy-community-qa/tree/master/recipes to test installing usegalaxy's toolset in a multi-process environment.

ahhhh, you mean not me manually doing this? Sure, seems like the perfect place to do that.

**Comment by martenson** [comment]
@mvdbeek yeah, this looks like a good test case of a quite complex thing that we want to work reliably

**Pull request merged**